File: java/com/purescraps/purehtml/transformers/types/LengthTransformer.java
package com.purescraps.purehtml.transformers.types;

import com.purescraps.purehtml.transformers.TransformParams;
import com.purescraps.purehtml.transformers.Transformer;

import java.util.List;

public class LengthTransformer extends Transformer {
    private List<String>  args;
    public LengthTransformer(List<String>  args) {
        this.args = args;
    }
    public LengthTransformer(){
        this.args = null;
    }

    public void setArgs(List<String> args){
        this.args = args;
    }
    // Static method to return the name of the transformer
    public String getName() {
        return "length";
    }

    // The transform method that returns the length of a string or an array
    public Integer transform(TransformParams params) {
        Object val = params.getVal();

        // If the value is null, return null
        if (val == null) {
            return null;
        }

        // If the value is a String or an array, return the length
        if (val instanceof String) {
            return ((String) val).length();
        } else if (val.getClass().isArray()) {
            return java.lang.reflect.Array.getLength(val);
        }

        // If the value is neither a String nor an array, throw an error
        System.out.println("Length.transform: invalid value type: " + val.getClass().getName());
        System.exit(-1);
        return 0;
    }
}

File: python/purehtml/transformers/types/length_transformer.py
from abc import ABC
from typing import Any
from typing import List

from purehtml.transformers.transformer import Transformer


class LengthTransformer(Transformer, ABC):
    def __init__(self, args: List[str] = None):
        """
        Initialize the transformer with optional arguments.
        :param args: Arguments to configure the transformer (unused in this case).
        """
        super().__init__()
        self.args = args if args is not None else []

    def set_args(self, args: List[str]):
        """
        Set the list of attributes to retrieve.
        :param args: A list of attribute names.
        """
        self.args = args

    def get_name(self) -> str:
        """
        Return the name of the transformer.
        :return: The name of the transformer.
        """
        return "length"

    def transform(self, params) -> Any:
        """
        Returns the length of the given value if it's a string or array.
        :param params: The parameters that contain the value to calculate length from.
        :return: The length of the string or array, or raises an exception for invalid types.
        """
        val = params.get_val()

        # If the value is None, return None
        if val is None:
            return None

        # If the value is a string, return its length
        if isinstance(val, str):
            return len(val)

        # If the value is a list (array), return its length
        elif isinstance(val, (list, tuple)):
            return len(val)

        # If the value is neither a string nor a list/tuple, throw an error
        raise ValueError(f"LengthTransformer --> Invalid value type: {type(val)}")
